### PR TITLE
browser(firefox): implement RemoteDebuggingPipe

### PR DESCRIPTION
**Pull request by dgozman**

### CODE REVIEW

File: browser_patches/firefox/juggler/components/juggler.js [hunk 1]
```diff
@@ -32,48 +32,75 @@ CommandLineHandler.prototype = {
...
+    } else if (jugglerPipeFlag) {
+      const pipe = Cc['@mozilla.org/juggler/remotedebuggingpipe;1'].getService(Ci.nsIRemoteDebuggingPipe);
+      const connection = {
+        QueryInterface: ChromeUtils.generateQI([Ci.nsIRemoteDebuggingPipeClient]),
+        receiveMessage(message) {
+          if (this.onmessage)
+            this.onmessage({ data: message });
+        },
+        send(message) {
+          dump(`Sending ${message}\n`);
```
Comment (yury-s) [inline browser_patches/firefox/juggler/components/juggler.js:89]: Revert logging?

File: browser_patches/firefox/juggler/pipe/nsRemoteDebuggingPipe.h [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+  static already_AddRefed<nsIRemoteDebuggingPipe> GetSingleton();
+  nsRemoteDebuggingPipe();
+
+ private:
+  RefPtr<nsIRemoteDebuggingPipeClient> mClient;
+  nsCOMPtr<nsIThread> mReaderThread;
+  nsCOMPtr<nsIThread> mWriterThread;
+  std::atomic<bool> m_terminated { false };
+
+  void ReaderLoop();
```
Comment (yury-s) [inline browser_patches/firefox/juggler/pipe/nsRemoteDebuggingPipe.h:28]: style nit: move method declarations above fields?

File: browser_patches/firefox/juggler/pipe/nsRemoteDebuggingPipe.cpp [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+    }
+    if (start != 0 && start < line.size()) {
+      memmove(line.data(), line.data() + start, line.size() - start);
+    }
+    line.resize(line.size() - start);
+  }
+}
+
+void nsRemoteDebuggingPipe::ReceiveMessage(const nsCString& aMessage) {
+  if (mClient) {
```
Comment (yury-s) [inline browser_patches/firefox/juggler/pipe/nsRemoteDebuggingPipe.cpp:180]: nit: add MOZ_RELEASE_ASSERT(NS_IsMainThread() ?

**Review by yury-s — APPROVED** [review]

**Comment by dgozman** [comment]
All done.

**Pull request merged**